tools/create_service_bundle.py

```diff
@@ -0,0 +1,568 @@
+                    agent_id,
+                    task_file["path"],
+                    os.path.join(output_base_path, task_file_basename),
+                )
+
+
+def download_task_files(
+    agent_id: str,
+    executor_sandbox_path: str,
+    task_id: str,
+    base_path: str,
+    patterns_to_download: List[str] = [],
+) -> List[dict]:
+    executor_sandbox = browse_executor_sandbox(agent_id, executor_sandbox_path)
+    task_sandbox = browse_task_sandbox(agent_id, executor_sandbox_path, task_id)
+
+    # Pod task: download both its logs and its parent executor's logs.
+    if task_sandbox:
+        output_task_directory = os.path.join(base_path, task_id, "task")
+        download_sandbox_files(
+            agent_id, executor_sandbox, output_task_directory, patterns_to_download
+        )
+
+        output_executor_directory = os.path.join(base_path, task_id, "executor")
+        download_sandbox_files(
+            agent_id, task_sandbox, output_executor_directory, patterns_to_download
+        )
+    # Scheduler task: no parent executor, download only scheduler logs.
+    else:
+        output_directory = os.path.join(base_path, task_id)
+        download_sandbox_files(
+            agent_id, task_sandbox, output_directory, patterns_to_download
+        )
+
+
+class Bundle(object):
+    def __init__(self, package_name, service_name, directory_name):
+        self.package_name = package_name
+        self.service_name = service_name
+        self.directory_name = directory_name
+        self.framework_id = get_framework_id(self.package_name, self.service_name)
+
+    def tasks_with_state(self, state):
+        return list(
+            filter(
+                lambda task: task["state"] == state,
+                get_tasks(self.framework_id)["tasks"],
+            )
+        )
+
+    def running_tasks(self):
+        return self.tasks_with_state("TASK_RUNNING")
+
+    def tasks_with_state_and_prefix(self, state, prefix):
+        return list(
+            filter(
+                lambda task: task["name"].startswith(prefix),
+                self.tasks_with_state(state),
+            )
+        )
+
+    def write_file(self, file_name, content, serialize_to_json=False):
+        file_path = os.path.join(self.directory_name, file_name)
+
+        with open(file_path, "w") as f:
+            logger.info("Writing file {}".format(file_path))
+            if serialize_to_json:
+                json.dump(content, f, indent=2, sort_keys=True)
+            else:
+                f.write(content)
+                f.write("\n")
+
+    def run_on_tasks(self, fn, task_ids):
+        for task_id in task_ids:
+            fn(task_id)
+
+    def for_each_running_task(self, fn):
+        task_ids = list(map(lambda task: task["id"], self.running_tasks()))
+        self.run_on_tasks(fn, task_ids)
+
+    def for_each_running_task_with_prefix(self, prefix, fn):
+        task_ids = list(
+            map(
+                lambda task: task["id"],
```

Comment:
Use `itemgetter("id")` instead?

Reply:
Although the following would be much more readable IMHO:
```
task_ids = [task["id"] for task in self.running_tasks() if task["name"].startswith(prefix)]
```

Reply:
This is also more pythonic, and it may just be inexperience, but I find reading streaming code in Python always requires me to spend a couple of times reading it.

Reply:
This has been fixed.